File: lib/unmagic/components/password.rb
# frozen_string_literal: true

module Unmagic
  module Components
    # A password input with a button that shows what was typed. See
    # FormBuilder#password_field and ActionViewHelpers#password_field_tag.
    module Password
      # Wraps a rendered password input in <unmagic-password> with its toggle. The
      # button renders hidden and the element reveals it on upgrade, so without
      # script there is a working password field and no dead button.
      def self.wrap(view, input, id:)
        label = I18n.t("unmagic.components.password.show", default: "Show password")

        view.content_tag("unmagic-password", class: "UnmagicPassword") do
          view.safe_join [
            input,
            view.tag.button(type: "button", class: "UnmagicButton UnmagicButton--icon UnmagicPassword__toggle",
              "aria-controls": id, "aria-pressed": "false", "aria-label": label, title: label, hidden: true) do
              view.safe_join [
                Icons.svg(view, :eye, class: "UnmagicPassword__show"),
                Icons.svg(view, :eye_off, class: "UnmagicPassword__hide")
              ]
            end
          ]
        end
      end
    end
  end
end

File: pr-13/assets/javascripts/unmagic/components/password.js
// <unmagic-password> — a password input with a button that shows what was
// typed, rendered by password_field(reveal: true).
//
// The button renders hidden and is shown here, so without script there is a
// working password field and no dead button. It is always called "Show password",
// with aria-pressed while revealed. Toggling keeps the input's selection.

class UnmagicPassword extends HTMLElement {
  connectedCallback() {
    const button = this.querySelector(":scope > button")
    if (button) button.hidden = false
    this.addEventListener("click", this.#clicked)
  }

  disconnectedCallback() {
    this.removeEventListener("click", this.#clicked)
  }

  #clicked = (event) => {
    const button = event.target instanceof Element && event.target.closest("button")
    const input = this.querySelector("input")
    if (!button || !input) return

    const revealed = input.type === "password"
    const { selectionStart, selectionEnd } = input
    input.type = revealed ? "text" : "password"
    button.setAttribute("aria-pressed", String(revealed))
    try {
      input.setSelectionRange(selectionStart, selectionEnd)
    } catch {
      // Some inputs refuse a selection; the caret just moves to the end.
    }
  }
}

customElements.get("unmagic-password") || customElements.define("unmagic-password", UnmagicPassword)
